flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java

```diff
@@ -0,0 +1,380 @@
+
+    /** A list of listener per upload (~ per checkpoint plus pre-emptive uploads). */
+    @GuardedBy("lock")
+    private final List<UploadCompletionListener> uploadCompletionListeners = new ArrayList<>();
+
+    private final FsStateChangelogCleaner cleaner;
+
+    /** Current {@link SequenceNumber}. */
+    private SequenceNumber activeSequenceNumber = INITIAL_SQN;
+
+    /**
+     * {@link SequenceNumber} before which changes will NOT be requested, exclusive. Increased after
+     * materialization.
+     */
+    @GuardedBy("lock")
+    private SequenceNumber lowestSequenceNumber = INITIAL_SQN;
+
+    /** Active changes, that will all use {@link #activeSequenceNumber}. */
+    private List<StateChange> activeChangeSet = new ArrayList<>();
+
+    /** {@link #activeChangeSet} size in bytes. */
+    private long activeChangeSetSize;
+
+    /** Changes that are not yet uploaded (upload not requested). */
+    private final NavigableMap<SequenceNumber, StateChangeSet> notUploaded = new TreeMap<>();
+
+    /** Uploaded changes, ready for use in snapshots. */
+    @GuardedBy("lock")
+    private final NavigableMap<SequenceNumber, UploadResult> uploaded = new TreeMap<>();
+
+    /**
+     * Highest {@link SequenceNumber} for which upload has failed (won't be restarted), inclusive.
+     */
+    @Nullable
+    @GuardedBy("lock")
+    private Tuple2<SequenceNumber, Throwable> highestFailed;
+
+    @GuardedBy("lock")
+    private boolean closed;
+
+    FsStateChangelogWriter(
+            UUID logId,
+            KeyGroupRange keyGroupRange,
+            StateChangeUploader uploader,
+            long preEmptivePersistThresholdInBytes,
+            FsStateChangelogCleaner cleaner) {
+        this.logId = logId;
+        this.keyGroupRange = keyGroupRange;
+        this.uploader = uploader;
+        this.preEmptivePersistThresholdInBytes = preEmptivePersistThresholdInBytes;
+        this.cleaner = cleaner;
+    }
+
+    @Override
+    public void append(int keyGroup, byte[] value) throws IOException {
+        LOG.trace("append to {}: keyGroup={} {} bytes", logId, keyGroup, value.length);
+        checkState(!closed, "%s is closed", logId);
+        activeChangeSet.add(new StateChange(keyGroup, value));
+        activeChangeSetSize += value.length;
+        if (activeChangeSetSize >= preEmptivePersistThresholdInBytes) {
+            LOG.debug(
+                    "pre-emptively flush {}Mb of appended changes to the common store",
+                    activeChangeSetSize / 1024 / 1024);
+            persistInternal(notUploaded.isEmpty() ? activeSequenceNumber : notUploaded.firstKey());
+        }
+    }
+
+    @Override
+    public SequenceNumber initialSequenceNumber() {
+        return INITIAL_SQN;
+    }
+
+    @Override
+    public SequenceNumber lastAppendedSequenceNumber() {
+        LOG.trace("query {} sqn: {}", logId, activeSequenceNumber);
+        SequenceNumber tmp = activeSequenceNumber;
+        rollover();
```

Comment:
I remember we have been already discussing this `rollover()` in a method that sounds like a getter, but I forgot the rationale.
1. Can you remember me what's the reason behind it here?
2. At the very least maybe add a comment explaining it?
3. Maybe we can refactor the code to be more self-documenting without the need for the comment? Rename the method? Split it into two?

Reply:
So your question is why `rollover()` needs to be called here, right?

This is because the caller must be able to use the returned sequence number
to distinguish changes before and after this call. At the same time, we don't 
want to increment SQN on each append (to avoid too many objects and segments
in the resulting file).

I'll add a comment (unless you have a better idea of how to make it more clear).